mypy/checkexpr.py

```diff
@@ -1945,6 +1952,12 @@ def visit_op_expr(self, e: OpExpr) -> Type:
                     return self.strfrm_checker.check_str_interpolation(e.left, e.right)
         left_type = self.accept(e.left)
 
+        proper_left_type = get_proper_type(left_type)
+        if isinstance(proper_left_type, TupleType) and e.op == '+':
+            proper_right_type = get_proper_type(self.accept(e.right))
+            if isinstance(proper_right_type, TupleType):
+                return self.concat_tuples(proper_left_type, proper_right_type)
```

Comment:
There is a slight complication here: it's possible that the fallback type has a custom `__add__` override that does something different from what's defined in `tuple`. In that case we can't safely use this special casing. So there should be a check that checks whether the `__add__` method in the left partial fallback comes from `builtins.tuple`. If not, we use the old code path below instead of using `concat_tuples`. Let me know if you need more hints about how to do this.

Reply:
Isn't the same thing applicable for `__radd__` and right tuple?

Reply:
Yes, you are correct.

Reply:
Added the check.